src/LibVulkan.jl

```diff
@@ -12,19 +12,6 @@ else
     const libvulkan = "libvulkan"
 end
 
-libvulkan_handle = C_NULL
-
-function __init__()
-    libname = get(ENV, "JULIA_VULKAN_SDK_LIBNAME", "")
-    locations = [get(ENV, "JULIA_VULKAN_SDK_SEARCH_PATH", "")]
-    if isempty(libname)
-        libname = Libdl.find_library(["libvulkan", "vulkan", "vulkan-1", "libvulkan.so.1"], locations)
-    end
-    @assert libname != "" "cannot detect Vulkan SDK."
-    global libvulkan_handle = Libdl.dlopen(libname)
-    @assert libvulkan_handle != C_NULL "cannot dlopen libvulkan."
-end
```

Comment:
Was there any reason to have this? `libvulkan_handle` is never used. Did you explicitly use it from somewhere? Plus, I'm not sure `dlopen`ing is a good idea inside an `__init__` (but I may be wrong)

Reply:
> Was there any reason to have this? 

To make sure the library can actually be successfully loaded. 

> Plus, I'm not sure dlopening is a good idea inside an __init__ (but I may be wrong)

In fact, it's mandatory to do those initialization stuff(e.g. `dlopen`,`dlsym`) inside `__init__` for  precompilation,  please refer to [the doc](https://docs.julialang.org/en/v1/manual/modules/#Module-initialization-and-precompilation) for more info.

Reply:
So, if I understand correctly, there are three options for `libvulkan`:
1. Keep it a library name, but to be resolved at `__init__` time: `const libvulkan = Ref{String}("")`
2. Keep it a library name, resolved at precompilation time: no changes to this PR
3. Make it a library pointer that is obtained at `__init__` time: `const libvulkan = Ref{Ptr{Cvoid}}(0)`. I don't think we can `ccall` into that directly though, so we'd need to wrap `ccall`s with `dlsym` to get a function pointer.

Possibility 1. would require that we change the generated code to `ccall` into `libvulkan[]` instead of `libvulkan`. That way, a user may change its library paths between compile time and runtime, and can precompile without needing a valid library. I think this would be the best option, especially for portability, compared to 2.
Possibility 3. seems unnecessary and a lot of work. If `Libdl.find_library` succeeds at `__init__`, then `ccall`ing into whatever libname is returned by that will succeed as well.

Reply:
> In fact, it's mandatory to do those initialization stuff(e.g. `dlopen`,`dlsym`) inside `__init__` for precompilation
I am not sure this is necessary. All we need is a libname for `ccall` to work, right?

Reply:
> Possibility 1. would require that we change the generated code to ccall into libvulkan[] instead of libvulkan. That way, a user may change its library paths between compile time and runtime, and can precompile without needing a valid library.

Note that using non-constants `libvulkan[]` in `ccall` is not allowed in Julia-1.6-.

Reply:
The type stays constant, maybe that's enough. I tested it locally and for me it works well, and no crashes.

Reply:
Did you test it with `--compiled-modules=no`? If you tested it with this PR, it probably still used the precompiled value.

Reply:
I did it with `Pkg.test()`. I retried with `julia --project=. --compiled-modules=no -e 'using Pkg; Pkg.test()'` and it worked fine. Plus, the precompiled module is invalidated whenever the original module changes, and it was a fresh Julia session.

Reply:
Did you tested on Julia 1.5? I got the following error on Julia 1.5:
```
ERROR: TypeError: in ccall: first argument not a pointer or valid constant expression, expected Ptr, got a value of type Tuple{Symbol,String}
```

Reply:
Only Julia 1.6 for now. I guess if we want to support 1.5 we can't fetch the libname in `__init__`, therefore we'll have to keep it as it was (determined at compile time).
The downside is that it will likely not be portable once it's precompiled. So I think we should add a `@static if` branch to default to the old behavior for 1.5 and lower and keep the current one when supported.

Reply:
For some reason instance creation didn't succeed in CI with 1.5 though (for Linux, for the rest I don't know it errors before), so tests won't pass (with SwiftShader anyway). Should we:
1. Test only in 1.6 (and maybe highlight that in the README, while keeping the compat entry for 1.3 since it _may_ work but not guaranteed)?
2. Not test with SwiftShader for < 1.6 and take the SDK as before (but tests aren't really meaningful in that case)?
3. Other?

Reply:
[Here's](https://github.com/JuliaLang/julia/issues/29602#issuecomment-738074246) a related discussion about allowing non-constant libnames in `ccall`s Julia 1.6.

Reply:
Since the majority part of this package is auto-generated by Clang.jl, assuming there is no bug on the Clang.jl side, there shouldn't be any bugs here. Currently the only meaningful test of this package is probably for testing the `__init__` function. My two cents are 1. to leave the CI of VulkanCore.jl untouched; 2. to setup SwiftShader CI only for Vulkan.jl and do the meaningful tests there. I guess the following change on the VulkanCore.jl side would be enough for supporting SwiftShader. In this way, you need to follow [the instruction](](https://github.com/google/swiftshader#usage) ) and install SwiftShader to the system path or correctly setup some environment variables. 

```
@static if get(ENV, "JULIA_VULKAN_USE_SWIFTSHADER", "") != ""
     const libvulkan = "libvk_swiftshader"
 elseif Sys.iswindows()
     const libvulkan = "vulkan-1.dll" 
 elseif Sys.isapple()
     const libvulkan = "libvulkan.dylib"
 elseif Sys.islinux()
     const libvulkan = "libvulkan.so.1"
 else
     const libvulkan = "libvulkan"
end
```

Reply:
> assuming there is no bug on the Clang.jl side

For now, Clang.jl do have severe bugs, especially [this one](https://github.com/JuliaInterop/Clang.jl/issues/238), I'm doing an overhaul for the whole package now.

Reply:
Agreed for testing.
I still think that the behavior for finding libraries could be improved though. Is there a reason for splitting the libnames depending on the OS as you suggest? Can't we just try several default names and allow for a way to explicitly specify one like in this PR?
I think it would be better discussed in another PR that has only these changes. I'll make one for that, and one with the miscellaneous test fixes/refactoring.

Reply:
> [Here's](https://github.com/JuliaLang/julia/issues/29602#issuecomment-738074246) a related discussion about allowing non-constant libnames in `ccall`s Julia 1.6.

Anyway, Julia 1.6 has not been released yet. Let's wait for an official solution and use the workaround I mentioned above.

Reply:
> I think it would be better discussed in another PR that has only these changes. I'll make one for that, and one with the miscellaneous test fixes/refactoring.

Those test fixes look good to me. Let's merge them first. Thanks for the contribution!